utils/sorting.hh

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <stdexcept>
+#include <vector>
+#include <map>
+#include <seastar/core/future.hh>
+#include "seastar/coroutine/maybe_yield.hh"
+
+namespace utils {
+
+/**
+ * Topological sort a DAG using Kahn's algorithm.
+ *
+ * @param vertices      Contains all vertices of the graph
+ * @param adjecency_map Entry {k, v} means there is an k->v edge in the graph
+ * @return a vector of topologically sorted vertices
+ * @throws std::runtime_error when a graph has any cycle.
+ */
+template<typename T, typename Compare = std::less<T>>
+seastar::future<std::vector<T>> topological_sort(const std::vector<T>& vertices, const std::multimap<T, T, Compare>& adjecency_map) {
+    std::map<T, uint32_t, Compare> ref_count_map; // Contains counters how many edges point (referece) to a vertex
```

Comment:
If you introduce a new generic utility, add a test along with it.

Reply:
Do we need a multimap, or is an unordered_multimap sufficient?

Reply:
I guess if we're inheriting the implementation we can go along with what we had before.

Reply:
> Do we need a multimap, or is an unordered_multimap sufficient?

This implementation is more like the one that we originally planned to introduce. We are counting references to objects in the map, so it's intentional that this is a map.

Reply:
ok, so what about ordered vs unordered?

Reply:
(ah my original comment was for adjacency map, not ref_count_map)

Reply:
If we want the algorithm to behave deterministically then I suppose we should use the ordered map.

But then we should also make sure that the ordering of the items makes sense (e.g. we are using `T = entry*` for the old code - we should compare by name there)

Reply:
We can always pass a comparator